src/runtime_src/core/include/experimental/xrt_aie.h

```diff
@@ -113,6 +115,68 @@ extern "C" {
 
 #endif
 
+/**
+ * xrtAIEOpenExclusive() - Open an AIE context with exclusive mode
+ *
+ * @handle:          Handle to the device
+ * @xclbinUUID:      UUID of the xclbin withe specified AIE
+ *
+ * Return:          0 on success, or appropriate error number.
+ *
+ * There are three supported AIE context
+ * 1) exclusive: Can fully access AIE array. At any time, there can be only
+ *               one exclusive context. If an exclusive context is opened,
+ *               no other context can be opened.
+ * 2) primary:   Can fully access AIE array. At any time, there can be only
+ *               one primary context. If a primary context is opened, only
+ *               shared context can be opened by other process.
+ * 3) shared:    Can do non-disruptive acc on AIE (monitor, stateless
+ *               operation, etc.). There can be multiple shared context
+ *               at the same time.
+ *
+ * Note: If application does not call xrtAIEOpenXXX explicitly, by default,
+ *       it will try to acquire primary context when it tries to access AIE
+ *       array through XRT APIs.
+ */
+int
+xrtAIEOpenExclusive(xrtDeviceHandle handle, const xuid_t xclbinUUID);
+
+/**
+ * xrtAIEOpenPrimary() - Open an AIE context with primary mode
+ *
+ * @handle:          Handle to the device
+ * @xclbinUUID:      UUID of the xclbin withe specified AIE
+ *
+ * Return:          0 on success, or appropriate error number.
+ */
+int
+xrtAIEOpenPrimary(xrtDeviceHandle handle, const xuid_t xclbinUUID);
+
+/**
+ * xrtAIEOpenShared() - Open an AIE context with shared mode
+ *
+ * @handle:          Handle to the device
+ * @xclbinUUID:      UUID of the xclbin withe specified AIE
+ *
+ * Return:          0 on success, or appropriate error number.
+ */
+int
+xrtAIEOpenShared(xrtDeviceHandle handle, const xuid_t xclbinUUID);
+
+/**
+ * xrtAIEOpenPrimary() - Close AIE context
+ *
+ * @handle:          Handle to the device
+ * @xclbinUUID:      UUID of the xclbin withe specified AIE
+ *
+ * Return:          0 on success, or appropriate error number.
+ *
+ * Note: After close context, any access to AIE, we will try to acquire
+ *       primary context by default.
+ */
+int
+xrtAIECloseContext(xrtDeviceHandle handle, const xuid_t xclbinUUID);
```

Comment:
Is this really to close a context?     Isn't it more close of an aie kind of handle?  In other words, should xrtAIEOpenBlahBlah() return a handle, that should then be closed with xrtAIEClose(handle) ?  What is the C++ equivalent , once implemented?   Won't the C++ code create an xrt::aie::device (akin to a handle) which is automatic (reference counted) and when it goes out of scope the context is automatically closed unbeknownst to the user?

Reply:
In C APIs, AIE related operations are using device handle rather than AIE specific handle
xrtAIESyncBO(xrtDeviceHandle handle, xrtBufferHandle bohdl
xrtAIEResetArray(xrtDeviceHandle handle);

We can return a handle to xrtAIEOpen_context_, but the only place this handle will be used is xrtCloseContext. I am not sure if we want to introduce a handle here. It would have been more straightforward if we have used an AIE handler in xrt AIE operation, like we have xrtGraphOpen, xrtPLKernelOpen.

Reply:
I realize that xrt::aie::device is just adding additional APIs on top of xrt::device, but maybe it should be more than just API extensions?    And maybe xrtAIEDeviceOpen should be corresponding to creating xrt::aie::device so that both work with same aie specific handle that can manage the context for the user?   It's odd that in C++ the user has to construct an xrt::aie::device but that in C there is no corresponding xrtAIEDeviceOpen().

Reply:
I made the related change based on our discussion. The following APIs are removed
xrtAIEOpenExcluseiv
xrtAIEOpenPrimary
xrtAIEOpenShared
xrtAIECloseContext

The following APIs are added
xrtAIEOpenDevice
xrtAIEOpenDeviceExclusive
xrtAIEOpenDeviceShared

Please take another look. Thanks